File: js/src/maplestory/api/dto/character/characterAndroidEquipmentDto.ts
import { CharacterAndroidCashItemEquipmentDto } from './characterAndroidCashItemEquipmentDto';
import { CharacterAndroidEquipmentFaceDto } from './characterAndroidEquipmentFaceDto';
import { CharacterAndroidEquipmentHairDto } from './characterAndroidEquipmentHairDto';
import { CharacterAndroidEquipmentPresetDto } from './characterAndroidEquipmentPresetDto'
import { CharacterAndroidEquipmentDtoBody } from '../../response/character/characterAndroidEquipmentDtoBody';

/**
 * 캐릭터 안드로이드 장비 정보
 */
class CharacterAndroidEquipmentDto {
  /**
   * 조회 기준일
   */
  date: Date;

  /**
   * 안드로이드 명
   */
  androidName: string | null;

  /**
   * 안드로이드 닉네임
   */
  androidNickname: string | null;

  /**
   * 안드로이드 아이콘
   */
  androidIcon: string | null;

  /**
   * 안드로이드 아이템 설명
   */
  androidDescription: string | null;

  /**
   * 안드로이드 헤어 정보
   */
  androidHair: CharacterAndroidEquipmentHairDto;

  /**
   * 안드로이드 성형 정보
   */
  androidFace: CharacterAndroidEquipmentFaceDto;

  /**
   * 안드로이드 피부 명
   */
  androidSkinName: string | null;

  /**
   * 안드로이드 캐시 아이템 장착 정보
   */
  androidCashItemEquipment: CharacterAndroidCashItemEquipmentDto[];

  /**
   * 안드로이드 이어센서 클립 적용 여부
   */
  androidEarSensorClipFlag: string | null;

  /**
   * 안드로이드 성별
   */
  androidGender: string | null;

  /**
   * 안드로이드 등급
   */
  androidGrade: string | null;

  /**
   * 비인간형 안드로이드 여부
   */
  androidNonHumanoidFlag: string | null;

  /**
   * 잡화상점 기능 이용 가능 여부
   */
  androidShopUsableFlag: string | null;

  /**
   * 적용 중인 장비 프리셋 번호
   */
  presetNo: number | null;

  /**
   * 1번 프리셋 안드로이드 정보
   */
  androidPreset1: CharacterAndroidEquipmentPresetDto | null;

  /**
   * 2번 프리셋 안드로이드 정보
   */
  androidPreset2: CharacterAndroidEquipmentPresetDto | null;

  /**
   * 3번 프리셋 안드로이드 정보
   */
  androidPreset3: CharacterAndroidEquipmentPresetDto | null;

  constructor(obj: CharacterAndroidEquipmentDtoBody) {
    const {
      date,
      android_name,
      android_nickname,
      android_icon,
      android_description,
      android_hair,
      android_face,
      android_skin_name,
      android_cash_item_equipment,
      android_ear_sensor_clip_flag,
      android_gender,
      android_grade,
      android_non_humanoid_flag,
      android_shop_usable_flag,
      preset_no,
      android_preset_1,
      android_preset_2,
      android_preset_3,
    } = obj;

    this.date = new Date(date);
    this.androidName = android_name;
    this.androidNickname = android_nickname;
    this.androidIcon = android_icon;
    this.androidDescription = android_description;
    this.androidHair = new CharacterAndroidEquipmentHairDto(android_hair);
    this.androidFace = new CharacterAndroidEquipmentFaceDto(android_face);
    this.androidSkinName = android_skin_name;
    this.androidCashItemEquipment = android_cash_item_equipment.map(
      (equipment) => new CharacterAndroidCashItemEquipmentDto(equipment),
    );
    this.androidEarSensorClipFlag = android_ear_sensor_clip_flag ?? null;
    this.androidGender = android_gender;
    this.androidGrade = android_grade;
    this.androidNonHumanoidFlag = android_non_humanoid_flag;
    this.androidShopUsableFlag = android_shop_usable_flag;
    this.presetNo = preset_no;
    this.androidPreset1 = android_preset_1 ? new CharacterAndroidEquipmentPresetDto(android_preset_1) : null;
    this.androidPreset2 = android_preset_2 ? new CharacterAndroidEquipmentPresetDto(android_preset_2) : null;
    this.androidPreset3 = android_preset_3 ? new CharacterAndroidEquipmentPresetDto(android_preset_3) : null;
  }
}

export { CharacterAndroidEquipmentDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/msea/dto/character/CharacterAndroidEquipmentDTO.java
package dev.spiralmoon.maplestory.api.msea.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import dev.spiralmoon.maplestory.api.Utils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.ZonedDateTime;
import java.util.List;

/**
 * Character android equipment information
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterAndroidEquipmentDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterAndroidEquipmentDTO<
        CharacterAndroidEquipmentHairDTO,
        CharacterAndroidEquipmentFaceDTO,
        CharacterAndroidEquipmentSkinDTO,
        CharacterAndroidCashItemEquipmentDTO,
        CharacterAndroidEquipmentPresetDTO> {

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    @JsonProperty("date")
    private String date;

    /**
     * Android name
     */
    @JsonProperty("android_name")
    private String androidName;

    /**
     * Android nickname
     */
    @JsonProperty("android_nickname")
    private String androidNickname;

    /**
     * Android icon
     */
    @JsonProperty("android_icon")
    private String androidIcon;

    /**
     * Android item description
     */
    @JsonProperty("android_description")
    private String androidDescription;

    /**
     * Android hair information
     */
    @JsonProperty("android_hair")
    private CharacterAndroidEquipmentHairDTO androidHair;

    /**
     * Android face information
     */
    @JsonProperty("android_face")
    private CharacterAndroidEquipmentFaceDTO androidFace;

    /**
     * Android skin information
     */
    @JsonProperty("android_skin")
    private CharacterAndroidEquipmentSkinDTO androidSkin;

    /**
     * Android cash item equipment information
     */
    @JsonProperty("android_cash_item_equipment")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterAndroidCashItemEquipmentDTO> androidCashItemEquipment;

    /**
     * Whether the Android ear sensor clip is applied
     */
    @JsonProperty("android_ear_sensor_clip_flag")
    private String androidEarSensorClipFlag;

    /**
     * Android gender
     */
    @JsonProperty("android_gender")
    private String androidGender;

    /**
     * Android grade
     */
    @JsonProperty("android_grade")
    private String androidGrade;

    /**
     * Whether the Android is non-humanoid
     */
    @JsonProperty("android_non_humanoid_flag")
    private String androidNonHumanoidFlag;

    /**
     * Whether the Android can use the general store functionality
     */
    @JsonProperty("android_shop_usable_flag")
    private String androidShopUsableFlag;

    /**
     * Active equipment preset number
     */
    @JsonProperty("preset_no")
    private Integer presetNo;

    /**
     * Android information for preset 1
     */
    @JsonProperty("android_preset_1")
    private CharacterAndroidEquipmentPresetDTO androidPreset1;

    /**
     * Android information for preset 2
     */
    @JsonProperty("android_preset_2")
    private CharacterAndroidEquipmentPresetDTO androidPreset2;

    /**
     * Android information for preset 3
     */
    @JsonProperty("android_preset_3")
    private CharacterAndroidEquipmentPresetDTO androidPreset3;

    /**
     * Reference date for query (SGT, daily data with hours and minutes set to 0)
     */
    public ZonedDateTime getDate() {
        return date != null ? Utils.toZonedDateTime(this.date) : null;
    }
}
